Advertise image optimization

daniloercoli opened the pull request:
Fixes #5572 by showing a dialog when the user taps on `Add media` button in the toolbar.

This image optimization promo dialog is shown on the screen when all of the following conditions are met:
1. Image optimization is OFF on the site.
2. Didn't already ask to enable the feature.
3. <del>The user has granted storage access to the app. *This is because we don't want to ask so much things to users the first time they try to add a picture to the app.*</del>

cc @nbradbury Any thoughts on this? 
Do you think the button is the right place to show this promo dialog?

nbradbury commented:
I think it would make more sense to show this after the user has selected a photo instead of right after the photo picker is shown.

Also, I think this is a bit wordy:

![screenshot_1498932323](https://user-images.githubusercontent.com/3903757/27764555-353cf26c-5e6a-11e7-809c-2f925c820e14.png)

How about:

![screenshot_1498933891](https://user-images.githubusercontent.com/3903757/27764556-3a3835f6-5e6a-11e7-9366-aa0fa25e1abc.png)

nbradbury commented:
Do you think we also need this in the media browser, when the user chooses to add a photo to their media library?

daniloercoli commented:
> Do you think we also need this in the media browser, when the user chooses to add a photo to their media library?

Thanks for your feedback @nbradbury - I've added it to the Media browser. Also changed the logic a bit, and now `Advertise Images` is only shown when you took a new picture. (After the picture is taken, and selected to be inserted in the post). It was a kind of weird asking upfront.

Could not add it to the other flows without making a lot of changes to the current code. So I skipped it for now.

Also changed the wording according to your comment above.

nbradbury commented:
Instead of NO / YES, I think the buttons work better as LEAVE OFF / TURN ON as suggested [earlier](https://github.com/wordpress-mobile/WordPress-Android/pull/6186#issuecomment-312448490).

nbradbury commented:
Just so I understand, for now we're only showing the dialog when the user takes a photo? I think it should also appear when the user adds a photo from their device, but that can be tackled separately if necessary.

nbradbury commented:
When this dialog is shown from the media browser, rotating the device loses the dialog and the picture is never added. This doesn't happen in `EditPostActivity` because that activity has `android:configChanges="orientation|keyboardHidden|screenSize"` in the manifest.

daniloercoli commented:
> Instead of NO / YES, I think the buttons work better as LEAVE OFF / TURN ON as suggested earlier.

Oh sorry, I missed them. Changed in 1902464.

daniloercoli commented:
> should also appear when the user adds a photo from their device

I initially thought to keep this case off, but yes it makes sense to advertise images optimization for picture from device's library. Added in 1dc4d38

daniloercoli commented:
> Fixing this will likely require a DialogFragment, which is going to complicate things here.

Yes, tried to fix the issue but things were really getting too complicated, and code looks goofy. 
Too much changes are required in the 2 activities in order to use a DialogFragment for a simple thing like this. 
*Removed the Ads in MediaLibrary for now. The Promo is only shown in EditPost.*

Ready for another pass!

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -585,8 +585,34 @@ private void hidePhotoPicker() {
+        if (WPMediaUtils.shouldAdvertiseImageOptimization(this, mSite)) {
+            boolean hasSelectedPicture = false;
+            for (Uri uri : uriList) {
+                if (!MediaUtils.isVideo(uri.toString())) {
```
This is an edge case, but it's possible the user selected a document or audio file. So this should instead check whether it's a valid image.

nbradbury commented:
Looks good! :shipit:
